src/typescript-service.ts

```diff
@@ -83,13 +84,161 @@ export default class TypeScriptService {
+                    importDirs.add(impDir);
+                }
+
+                return Promise.all(
+                    Array.from(importDirs).map((impDir) => {
+                        const filesToEnsure = []
+                        return this.projectManager.walkRemote(impDir, function(path: string, info: FileSystem.FileInfo, err?: Error): (Error | null) {
+                            if (err) {
+                                return err ? err : null;
+                            } else if (info.dir) {
+                                return null;
+                            }
+
+                            if (util.isJSTSFile(info.name)) {
+                                filesToEnsure.push(info.name);
+                            }
+
+                            return null;
+                        }).then(() => {
+                            return this.projectManager.ensureFiles(filesToEnsure)
+                        });
+                    })
+                );
+            });
+        }).then(() => {
+            this.projectManager.refreshConfigurations();
+        });
+
+        this.ensuredFilesForHoverAndDefinition.set(uri, promise);
+        promise.catch((err) => {
+            console.error("Failed to fetch files for hover/definition for uri ", uri, ", error:", err);
+            this.ensuredFilesForHoverAndDefinition.delete(uri);
+        });
+        return promise;
+    }
+
+    private ensuredFilesForWorkspaceSymbol: Promise<void> = null;
+
+    private ensureFilesForWorkspaceSymbol(): Promise<void> {
+        if (this.ensuredFilesForWorkspaceSymbol) {
+            return this.ensuredFilesForWorkspaceSymbol;
+        }
+
+        const self = this;
+        const filesToEnsure = [];
+        const promise = this.projectManager.walkRemote(this.projectManager.getRemoteRoot(), function(path: string, info: FileSystem.FileInfo, err?: Error): (Error | null) {
+            if (err) {
+                return err ? err : null;
+            } else if (info.dir) {
+                if (info.name.indexOf("/node_modules/") !== -1) {
+                    return pm.skipDir;
+                } else {
+                    return null;
+                }
+            }
+            if (util.isJSTSFile(path)) {
+                filesToEnsure.push(path);
+            }
+            return null;
+        }).then(() => {
+            this.projectManager.ensureFiles(filesToEnsure)
+        }).then(() => {
+            this.projectManager.refreshConfigurations();
+        });
+
+        this.ensuredFilesForWorkspaceSymbol = promise;
+        promise.catch((err) => {
+            console.error("Failed to fetch files for workspace/symbol:", err);
+            this.ensuredFilesForWorkspaceSymbol = null;
+        });
+
+        return promise;
+    }
+
+    private ensuredAllFiles: Promise<void> = null;
+
+    private ensureFilesForReferences(uri: string): Promise<void> {
```

Comment:
There is one difference. `ensureFilesForReferences` fetches all files, including those in `node_modules` if the referenced token is in a `node_modules` file. It does so, because references to those tokens could be defined in any file in the repository. It avoids doing this when it doesn't need to, because `node_modules` is often huge.